File: python/gink/impl/sequence.py
from typing import Optional, Iterable, Union, Tuple
from typeguard import typechecked
from random import randint

# gink implementation
from .builders import ChangeBuilder
from .typedefs import GenericTimestamp, MuTimestamp, UserValue
from .container import Container
from .muid import Muid
from .database import Database
from .bundler import Bundler
from .coding import SEQUENCE
from .tuples import PositionedEntry, SequenceKey
from .utilities import generate_timestamp


class Sequence(Container):
    BEHAVIOR = SEQUENCE

    @typechecked
    def __init__(
            self,
            muid: Optional[Union[Muid, str]] = None,
            *,
            arche: Optional[bool] = None,
            contents: Optional[Iterable[Union[UserValue, Container]]] = None,
            database: Optional[Database] = None,
            bundler: Optional[Bundler] = None,
            comment: Optional[str] = None,
    ):
        """
        Constructor for a sequence proxy.

        muid: the global id of this container, created on the fly if None
        arche: whether this will be the global version of this container (accessible by all databases)
        contents: prefill the sequence with an iterable of values upon initialization
        database: database send bundles through, or last db instance created if None
        bundler: the bundler to add changes to, or a new one if None and immediately commits
        comment: optional comment to add to the bundler
        """
        immediate = False
        if bundler is None:
            immediate = True
            bundler = Bundler(comment)

        Container.__init__(
                self,
                behavior=SEQUENCE,
                muid=muid,
                arche=arche,
                database=database,
                bundler=bundler,
        )
        if contents is not None:
            self.clear(bundler=bundler)
            self.extend(contents, bundler=bundler)
        if immediate and len(bundler):
            self._database.bundle(bundler)

    def __iter__(self):
        for thing in self.values():
            yield thing

    def dumps(self, as_of: GenericTimestamp = None) -> str:
        if self._muid.medallion == -1 and self._muid.timestamp == -1:
            identifier = "arche=True"
        else:
            identifier = f"muid={self._muid!r}"
        result = f"""{self.__class__.__name__}({identifier}, contents=["""
        stuffing = [repr(val) for val in self.values(as_of=as_of)]
        as_one_line = result + ", ".join(stuffing) + "])"
        if len(as_one_line) < 80:
            return as_one_line
        result += "\n\t"
        result += ",\n\t".join(stuffing) + "])"
        return result

    @typechecked
    def append(
            self,
            value: Union[UserValue, Container], *,
            expiry: GenericTimestamp = None,
            bundler=None,
            comment=None
    ) -> Muid:
        """ Append value to the end of the queue.

            If expiry is set, the added entry will be removed at the specified time.
        """
        return self._add_entry(value=value, bundler=bundler, comment=comment, expiry=expiry)

    @typechecked
    def insert(
            self,
            index: int,
            value: Union[UserValue, Container],
            expiry: GenericTimestamp = None,
            bundler=None,
            comment=None
    ) -> Muid:
        """ Inserts value before index.

            The resulting entry expires at expiry time if specified, which must be in the future.

            If no bundler is passed, applies the changes immediately, with comment.
            Otherwise, just appends the necessary changes to the passed bundler.

            returns the muid of the entry
        """
        return self._add_entry(
            value=value,
            effective=self._position(before=index),
            bundler=bundler,
            comment=comment,
            expiry=expiry)

    @typechecked
    def extend(
            self,
            iterable: Iterable[Union[UserValue, Container]], *,
            expiries: Union[GenericTimestamp, Iterable[GenericTimestamp]] = None,
            bundler=None,
            comment=None,
    ):
        """ Adds all of the items in iterable at once to this sequence.

            expiries, if present, may be either a single expiry to be applied to all new entries,
            or a iterable of expiries of the same length as the data

            Since all items will be appended to the sequence in the same transaction, they will
            all have the same timestamp, and so it won't be possible to move anything between them.

            returns the bundler (either passed or created on the fly)
        """
        immediate = False
        if not isinstance(bundler, Bundler):
            immediate = True
            bundler = Bundler(comment)
        items = list(iterable)
        if hasattr(expiries, "__iter__"):
            expiries = list(expiries)  # type: ignore
        for i in range(len(items)):
            if isinstance(expiries, list):
                expiry = expiries[i]
            else:
                expiry = expiries
            expiry = self._database.resolve_timestamp(expiry) if expiry else None  # type: ignore
            self._add_entry(value=items[i], bundler=bundler, expiry=expiry)
        if immediate and len(bundler):
            self._database.bundle(bundler)
        return bundler

    @typechecked
    def yank(self, muid: Muid, *, dest: GenericTimestamp = None, bundler=None, comment=None) -> Muid:
        """ Removes or moves an entry by muid.

            muid: what to move
            bundler: what to add this change to
            comment: make an immediate change with this comment
            dest: new location in the list or time in the future; integer values are interpreted
            to be before the given index if positive or after the given index if negative

            returns: the muid of the change
        """
        immediate = False
        if not isinstance(bundler, Bundler):
            immediate = True
            bundler = Bundler(comment)
        change_builder = ChangeBuilder()
        movement_builder = change_builder.movement  # type: ignore
        self._muid.put_into(movement_builder.container)
        muid.put_into(movement_builder.entry)
        if dest == -1:
            dest = generate_timestamp()
        elif isinstance(dest, int) and dest < 1e15:
            dest = self._position(before=dest) if dest >= 0 else self._position(after=dest)
        elif dest is None:
            dest = 0
        else:
            dest = self._database.resolve_timestamp(dest)
        assert isinstance(dest, int)
        movement_builder.dest = dest
        muid = bundler.add_change(change_builder)
        if immediate:
            self._database.bundle(bundler)
        return muid

    @typechecked
    def pop(self, index: int = -1, *, dest: GenericTimestamp = None, bundler=None, comment=None):
        """ (Re)move and return an item at index (default last).

            If nothing exists at the specified index will raise an IndexError.
            If bundler is specified, simply adds the change to that, otherwise applies it.
            If comment is specified and no bundler then will make change with that comment.

            If dest is specified, it may be a time to hid the entry until, or a time in the past
            to reposition the entry to (the list is ordered by timestamps).
        """
        sequence_key, entry_value = self.at(index)
        self.yank(sequence_key.entry_muid, dest=dest, bundler=bundler, comment=comment)
        return entry_value

    @typechecked
    def remove(
        self,
        value: Union[UserValue, Container], *,
        dest: GenericTimestamp = None,
        bundler=None,
        comment=None
    ) -> Muid:
        """ Remove first occurance of value.

            Raises ValueError if the value is not present.
        """
        as_of = self._database.resolve_timestamp()
        for positioned in self._database.get_store().get_ordered_entries(self._muid, as_of):
            found = self._get_occupant(positioned.builder, positioned.entry_muid)
            if found == value:
                return self.yank(positioned.entry_muid, dest=dest,
                                 bundler=bundler, comment=comment)
        raise ValueError("matching item not found")

    def items(self, *, as_of: GenericTimestamp = None) -> Iterable[Tuple[SequenceKey, Union[UserValue, Container]]]:
        """ Returns pairs of (muid, contents) for the sequence at the given time.
        """
        as_of = self._database.resolve_timestamp(as_of)
        for positioned in self._database.get_store().get_ordered_entries(self._muid, as_of=as_of):
            found = self._get_occupant(positioned.builder, positioned.entry_muid)
            sequence_key = SequenceKey(positioned.position, positioned.entry_muid)
            yield sequence_key, found

    def keys(self, *, as_of: GenericTimestamp = None) -> Iterable[SequenceKey]:
        for key, _ in self.items(as_of=as_of):
            yield key

    def values(self, *, as_of: GenericTimestamp = None) -> Iterable[Union[UserValue, Container]]:
        for _, val in self.items(as_of=as_of):
            yield val

    def __getitem__(self, what):
        """ Gets the specified item, either index counting up from
            zero, or negative number when counting from end,
            or whatever is found at an address in case of muid.
        """
        return self.at(what)[1]

    @typechecked
    def at(self, index: int, *, as_of: GenericTimestamp = None):
        """ Returns the ((position-ts, entry-muid), value) at the specified index.

            Index may be negative, in which case starts looking at the end.
            Raises IndexError if not present.
        """
        as_of = self._database.resolve_timestamp(as_of)
        offset = ~index if index < 0 else index
        iterable = self._database.get_store().get_ordered_entries(
            self._muid, as_of=as_of, limit=1, offset=offset, desc=(index < 0))
        for positioned in iterable:
            assert isinstance(positioned, PositionedEntry)
            found = self._get_occupant(positioned.builder, positioned.entry_muid)
            sequence_key = SequenceKey(positioned.position, positioned.entry_muid)
            return sequence_key, found
        raise IndexError(f"could not find anything at index {index}")

    def size(self, *, as_of: GenericTimestamp = None) -> int:
        """ Tells the size at the specified as_of time.
        """
        as_of = self._database.resolve_timestamp(as_of)
        count = 0
        for _ in self._database.get_store().get_ordered_entries(self._muid, as_of=as_of):
            count += 1
        return count

    @typechecked
    def index(self, value: Union[UserValue, Container], start=0, stop=None, *, as_of: GenericTimestamp = None) -> int:
        """ Return the first index of the value at the given time (or now).

            Raises a ValueError if the value isn't present and raise_if_missing is True,
            otherwise just returns None.
        """
        as_of = self._database.resolve_timestamp(as_of)
        index = start
        iterable = self._database.get_store().get_ordered_entries(self._muid, as_of, offset=start)
        for positioned in iterable:
            found = self._get_occupant(positioned.builder, positioned.entry_muid)
            if found == value:
                return index
            if stop is not None and index >= stop:
                break
            index += 1
        raise ValueError("matching item not found")

    @typechecked
    def __contains__(self, item: Union[UserValue, Container]) -> bool:
        """ Returns true if something matching item is in queue. """
        try:
            self.index(item)
            return True
        except ValueError:
            return False

    def _position(self, after: Optional[int] = None, before: Optional[int] = None) -> MuTimestamp:
        """ Gets a new position after or before the index of an existing entry. """
        if after is None and before is None:
            raise ValueError("need to specify at least one index")
        if before is None:
            assert after is not None
            if after >= 0:
                try:
                    self.at(after + 1)
                    before = after + 1
                except IndexError:
                    after = -1
        if before == 0:
            assert after is None
            position = self.at(0)[0][0]
            return position - randint(0, int(1e6))
        if after == -1:
            assert before is None
            position = self.at(-1)[0][0]
            return position + randint(0, int(1e3))
        if after is None:
            assert before is not None
            after = before - 1
        if before is None:
            assert after is not None
            before = after + 1
        at1 = self.at(after)
        at2 = self.at(before)
        p1 = at1[0].position
        p2 = at2[0].position
        assert isinstance(p1, int)
        assert isinstance(p2, int)
        if p1 == p2:
            raise ValueError("positions in terms of time are equal")
        if p2 < p1:
            p1, p2 = p2, p1
        if p2 - p1 < 2:
            raise ValueError("not enough space between them")
        return randint(p1 + 1, p2 - 1)


Database.register_container_type(Sequence)

File: javascript/implementation/Sequence.ts
import { Database } from "./Database";
import { Container } from "./Container";
import { AsOf, Entry, Muid, Value } from "./typedefs";
import { Bundler } from "./Bundler";
import {
    ensure,
    generateTimestamp,
    muidToBuilder,
    muidToString,
    muidTupleToMuid,
    muidTupleToString,
    wrapKey,
    wrapValue,
} from "./utils";
import { construct, interpret, toJson } from "./factories";
import { Behavior, ChangeBuilder, ContainerBuilder } from "./builders";
import { EntryBuilder } from "./builders";
import { movementHelper } from "./store_utils";

/**
 * Kind of like the Gink version of a Javascript Array; supports push, pop, shift.
 * Doesn't support unshift because order is defined by insertion order.
 */
export class Sequence extends Container {
    constructor(
        database: Database,
        address?: Muid,
        containerBuilder?: ContainerBuilder
    ) {
        super(database, address, Behavior.SEQUENCE);
        if (this.address.timestamp < 0) {
            //TODO(https://github.com/google/gink/issues/64): document default magic containers
            ensure(
                address.offset === Behavior.SEQUENCE,
                "magic tag not SEQUENCE"
            );
        } else {
            ensure(
                containerBuilder.getBehavior() === Behavior.SEQUENCE,
                "container not sequence"
            );
        }
    }

    /**
     * Adds an element to the end of the list.
     * @param value
     * @param change change set to apply the change to or comment to put in
     * @returns
     */
    async push(
        value: Value | Container,
        change?: Bundler | string
    ): Promise<Muid> {
        return await this.addEntry(undefined, value, change);
    }

    async move(
        muidOrPosition: Muid | number,
        dest: number,
        purge?: boolean,
        bundlerOrComment?: Bundler | string
    ) {
        let immediate = false;
        let bundler: Bundler;
        if (bundlerOrComment instanceof Bundler) {
            bundler = bundlerOrComment;
        } else {
            immediate = true;
            bundler = new Bundler(bundlerOrComment);
        }
        const store = this.database.store;
        // TODO: clarify what's going on here
        const muid =
            typeof muidOrPosition === "object"
                ? muidOrPosition
                : muidTupleToMuid(
                      Array.from(
                          (
                              await store.getOrderedEntries(
                                  this.address,
                                  muidOrPosition
                              )
                          ).values()
                      ).pop().entryId
                  );
        ensure(muid.timestamp && muid.medallion && muid.offset);
        await movementHelper(
            bundler,
            muid,
            this.address,
            await this.findDest(dest),
            purge
        );
        if (immediate) {
            await this.database.addBundler(bundler);
        }
    }

    async reset(args?: {
        toTime?: AsOf;
        bundlerOrComment?: Bundler | string;
        skipProperties?: boolean;
        recurse?: boolean;
        seen?: Set<string>;
    }): Promise<void> {
        const toTime = args?.toTime;
        const bundlerOrComment = args?.bundlerOrComment;
        const skipProperties = args?.skipProperties;
        const recurse = args?.recurse;
        const seen = recurse ? (args?.seen ?? new Set()) : undefined;
        if (seen) {
            seen.add(muidToString(this.address));
        }
        let immediate = false;
        let bundler: Bundler;
        if (bundlerOrComment instanceof Bundler) {
            bundler = bundlerOrComment;
        } else {
            immediate = true;
            bundler = new Bundler(bundlerOrComment);
        }
        if (!toTime) {
            // If no time is specified, we are resetting to epoch, which is just a clear
            this.clear(false, bundler);
        } else {
            const entriesThen = await this.database.store.getOrderedEntries(
                this.address,
                Infinity,
                toTime
            );
            // Need something subscriptable to compare by position
            const entriesNow = await this.database.store.getOrderedEntries(
                this.address,
                Infinity
            );

            for (const [key, entry] of entriesThen) {
                const placementTupleThen = entry.placementId;
                const placementNow = await this.database.store.getLocation(
                    muidTupleToMuid(entry.entryId)
                );
                const placementTupleNow = placementNow
                    ? placementNow.placement
                    : undefined;

                if (!placementNow) {
                    // This entry existed then, but has since been deleted
                    // Need to re-add it to the previous location
                    const entryBuilder = new EntryBuilder();
                    entryBuilder.setContainer(muidToBuilder(this.address));
                    entryBuilder.setKey(wrapKey(placementTupleThen[0]));
                    entryBuilder.setBehavior(entry.behavior);
                    if (entry.value !== undefined) {
                        entryBuilder.setValue(wrapValue(entry.value));
                    }

                    if (entry.pointeeList.length > 0) {
                        const pointeeMuid = muidTupleToMuid(
                            entry.pointeeList[0]
                        );
                        entryBuilder.setPointee(muidToBuilder(pointeeMuid));
                    }
                    const changeBuilder = new ChangeBuilder();
                    changeBuilder.setEntry(entryBuilder);
                    bundler.addChange(changeBuilder);
                } else {
                    if (
                        placementTupleNow &&
                        placementTupleThen[0] !== placementTupleNow[0]
                    ) {
                        // This entry exists, but has been moved
                        // Need to move it back
                        await movementHelper(
                            bundler,
                            muidTupleToMuid(entry.entryId),
                            this.address,
                            placementTupleThen[0],
                            false
                        );
                    }
                    // Need to remove the current entry from entriesNow if
                    // 1) the entry exists but was moved, or 2) the entry is untouched
                    ensure(
                        entriesNow.delete(
                            `${placementTupleNow[0]},${muidTupleToString(entry.entryId)}`
                        ),
                        "entry not found in entriesNow"
                    );
                }
                // Finally, if the previous entry was a container, recusively reset it
                if (seen && entry.pointeeList.length > 0) {
                    const pointeeMuid = muidTupleToMuid(entry.pointeeList[0]);
                    if (!seen.has(muidToString(pointeeMuid))) {
                        const container = await construct(
                            this.database,
                            pointeeMuid
                        );
                        await container.reset({
                            toTime,
                            bundlerOrComment: bundler,
                            skipProperties,
                            recurse,
                            seen,
                        });
                    }
                }
            }
            // We will need to loop through the remaining entries in entriesNow
            // to delete them, since we know they weren't in the sequence at toTime
            for (const [key, entry] of entriesNow) {
                await movementHelper(
                    bundler,
                    muidTupleToMuid(entry.entryId),
                    this.address,
                    undefined,
                    false
                );
            }
        }
        if (!skipProperties) {
            await this.database.resetContainerProperties(this, toTime, bundler);
        }
        if (immediate) {
            await this.database.addBundler(bundler);
        }
    }

    private async findDest(dest: number): Promise<number> {
        if (dest === 0 || dest === -1) {
            const currentFrontOrBack = <number>(
                (await this.getEntryAt(dest)).storageKey
            );
            return (
                currentFrontOrBack -
                Math.sign(dest + 0.5) * Math.floor(1e3 * Math.random())
            );
        }
        if (dest > +1e6) return dest;
        if (dest < -1e6) return generateTimestamp() + dest;
        const entryMap = await this.database.store.getOrderedEntries(
            this.address,
            dest
        );
        const entryArray = Array.from(entryMap.entries());
        const a = entryArray[entryArray.length - 2];
        const b = entryArray[entryArray.length - 1];
        const aTs = Number.parseInt(a[0].split(",")[0]);
        const bTs = Number.parseInt(b[0].split(",")[0]);
        if (Math.abs(aTs - bTs) < 2)
            throw new Error("can't find space between entries");
        return Math.floor((aTs + bTs) / 2);
    }

    /**
     * Removes and returns the specified entry of the list (default last),
     * in the provided change set or immediately if no CS is supplied.
     * Returns undefined when called on an empty list (and no changes are made).
     * @param what - position or Muid, defaults to last
     * @param purge - If true, removes so data cannot be recovered with "asOf" query
     * @param bundlerOrComment
     */
    async pop(
        what?: Muid | number,
        purge?: boolean,
        bundlerOrComment?: Bundler | string
    ): Promise<Container | Value | undefined> {
        let immediate = false;
        let bundler: Bundler;
        if (bundlerOrComment instanceof Bundler) {
            bundler = bundlerOrComment;
        } else {
            immediate = true;
            bundler = new Bundler(bundlerOrComment);
        }
        let returning: Container | Value;
        let muid: Muid;
        if (what && typeof what === "object") {
            muid = what;
            const entry = await this.database.store.getEntryById(muid);
            if (!entry) return undefined;
            ensure(
                entry.entryId[0] === muid.timestamp &&
                    entry.entryId[2] === muid.offset
            );
            returning = await interpret(entry, this.database);
        } else {
            what = typeof what === "number" ? what : -1;
            // Should probably change the implementation to not copy all intermediate entries into memory.
            const entries = Array.from(
                (
                    await this.database.store.getOrderedEntries(
                        this.address,
                        what
                    )
                ).values()
            );
            if (entries.length === 0) return undefined;
            const entry = entries[entries.length - 1];
            returning = await interpret(entry, this.database);
            muid = muidTupleToMuid(entry.entryId);
        }
        await movementHelper(bundler, muid, this.address, undefined, purge);
        if (immediate) {
            await this.database.addBundler(bundler);
        }
        return returning;
    }

    /**
     * Alias for this.pop(0, purge, bundlerOrComment)
     */
    async shift(
        purge?: boolean,
        bundlerOrComment?: Bundler | string
    ): Promise<Container | Value | undefined> {
        return await this.pop(0, purge, bundlerOrComment);
    }

    /**
     * Adds multiple entries into this sequence.
     * NOTE: If you pass a bundler, all changes will share the same timestamp. This means you will
     * not be able to move new entries in between these (you may move these entries between one another).
     * Without a bundler, each item from the iterable will be committed separately, which will be costly,
     * but there won't be the same restrictions on moving.
     * @param iterable An iterable of stuff to add to the sequence.
     * @param bundlerOrComment A bundler or comment for these changes
     */
    async extend(
        iterable: Iterable<Value | Container>,
        bundlerOrComment?: Bundler | string
    ): Promise<void> {
        for (const value of iterable) {
            await this.push(value, bundlerOrComment);
        }
    }

    private async getEntryAt(
        position: number,
        asOf?: AsOf
    ): Promise<Entry | undefined> {
        //TODO add a store method to only return the entry at a given location
        const entries = await this.database.store.getOrderedEntries(
            this.address,
            position,
            asOf
        );
        if (entries.size === 0) return undefined;
        if (position >= 0 && position >= entries.size) return undefined;
        if (position < 0 && Math.abs(position) > entries.size) return undefined;
        let val: Entry;
        for (let found of entries.values()) {
            val = found;
        }
        return val;
    }

    /**
     *
     * @param position Index to look for the thing, negative counts from end, or muid of entry
     * @param asOf
     * @returns value at the position of the list, or undefined if list is too small
     */
    async at(
        position: number,
        asOf?: AsOf
    ): Promise<Container | Value | undefined> {
        if (typeof position === "number") {
            const entry = await this.getEntryAt(position, asOf);
            return await interpret(entry, this.database);
        }
        throw Error("unexpected");
    }

    /**
     * Dumps the contents of this list to a javascript array.
     * useful for debugging and could also be used to export data by walking the tree
     * @param through how many elements to get, positive starting from beginning, negative starting from end
     * @param asOf effective time to get the dump for: leave undefined to get data as of the present
     * @returns an array containing Values (e.g. numbers, strings) and Containers (e.g. other Lists, Boxes, Directories)
     */
    async toArray(
        through = Infinity,
        asOf?: AsOf
    ): Promise<(Container | Value)[]> {
        const thisList = this;
        const entries = await thisList.database.store.getOrderedEntries(
            thisList.address,
            through,
            asOf
        );
        const applied = Array.from(entries.values());
        return await Promise.all(
            applied.map(async function (
                entry: Entry
            ): Promise<Container | Value> {
                return await interpret(entry, thisList.database);
            })
        );
    }

    async size(asOf?: AsOf): Promise<number> {
        const entries = await this.database.store.getOrderedEntries(
            this.address,
            Infinity,
            asOf
        );
        return entries.size;
    }

    /**
     * Function to iterate over the contents of the List, showing the address of each entry (which can be used in pop).
     * @param through count of many things to iterate through, positive starting from front, negative for end
     * @param asOf effective time to get the contents for
     * @returns an async iterator across everything in the list, with values returned being pairs of Muid, (Value|Container),
     */
    entries(
        through = Infinity,
        asOf?: AsOf
    ): AsyncGenerator<[Muid, Value | Container], void, unknown> {
        const thisList = this;
        return (async function* () {
            // Note: I'm loading all entries memory despite using an async generator due to shitty IndexedDb
            // behavior of closing transactions when you await on something else.  Hopefully they'll fix that in
            // the future and I can improve this.  Alternative, it might make sense to hydrate everything in a single pass.
            const entries = await thisList.database.store.getOrderedEntries(
                thisList.address,
                through,
                asOf
            );
            for (const entry of entries) {
                const hydrated = await interpret(entry[1], thisList.database);
                yield [muidTupleToMuid(entry[1].entryId), hydrated];
            }
        })();
    }

    /**
     * Generates a JSON representation of the data in the list.
     * Mostly intended for demo/debug purposes.
     * @param indent true to pretty print
     * @param asOf effective time
     * @param seen (internal use only! This prevents cycles from breaking things)
     * @returns a JSON string
     */
    async toJson(
        indent: number | boolean = false,
        asOf?: AsOf,
        seen?: Set<string>
    ): Promise<string> {
        if (seen === undefined) seen = new Set();
        ensure(indent === false, "indent not implemented");
        const mySig = muidToString(this.address);
        if (seen.has(mySig)) return "null";
        seen.add(mySig);
        const asArray = await this.toArray(Infinity, asOf);
        let returning = "[";
        let first = true;
        for (const value of asArray) {
            if (first) {
                first = false;
            } else {
                returning += ",";
            }
            returning += await toJson(
                value,
                indent === false ? false : +indent + 1,
                asOf,
                seen
            );
        }
        returning += "]";
        return returning;
    }
}
